lib/services/watcher.go

```diff
@@ -0,0 +1,373 @@
+	// return without subscription established first,
+	// Code line 3 could execute and line 4 could miss event,
+	// wrapping up with out of sync replica.
+	// To avoid this, before doing fetch,
+	// cache process makes sure the connection is established
+	// by receiving init event first.
+	select {
+	case <-watcher.Done():
+		return trace.ConnectionProblem(watcher.Error(), "watcher is closed")
+	case <-refetchC:
+		p.Log.Debug("Triggering scheduled refetch.")
+		return nil
+	case <-p.ctx.Done():
+		return trace.ConnectionProblem(p.ctx.Err(), "context is closing")
+	case event := <-watcher.Events():
+		if event.Type != types.OpInit {
+			return trace.BadParameter("expected init event, got %v instead", event.Type)
+		}
+	}
+
+	if err := p.getResourcesAndUpdateCurrent(p.ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	p.retry.Reset()
+
+	for {
+		select {
+		case <-watcher.Done():
+			return trace.ConnectionProblem(watcher.Error(), "watcher is closed")
+		case <-refetchC:
+			p.Log.Debug("Triggering scheduled refetch.")
+			return nil
+		case <-p.ctx.Done():
+			return trace.ConnectionProblem(p.ctx.Err(), "context is closing")
+		case event := <-watcher.Events():
+			if err := p.processEventAndUpdateCurrent(p.ctx, event); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+}
+
+// ProxyWatcherConfig is a ProxyWatcher configuration.
+type ProxyWatcherConfig struct {
+	ResourceWatcherConfig
+	// ProxyGetter is used to directly fetch the list of active proxies.
+	ProxyGetter
+	// ProxiesC is a channel used by the watcher to push updates to the proxy
+	// set.  It receives a fresh list at startup and subsequently a list of all
+	// known proxies whenever an addition or deletion is detected.
+	ProxiesC chan []types.Server
+}
+
+// CheckAndSetDefaults checks parameters and sets default values.
+func (cfg *ProxyWatcherConfig) CheckAndSetDefaults() error {
+	if err := cfg.ResourceWatcherConfig.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if cfg.ProxyGetter == nil {
+		getter, ok := cfg.Client.(ProxyGetter)
+		if !ok {
+			return trace.BadParameter("missing parameter ProxyGetter and Client not usable as ProxyGetter")
+		}
+		cfg.ProxyGetter = getter
+	}
+	if cfg.ProxiesC == nil {
+		cfg.ProxiesC = make(chan []types.Server)
+	}
+	return nil
+}
+
+// NewProxyWatcher returns a new instance of ProxyWatcher.
+func NewProxyWatcher(cfg ProxyWatcherConfig) (*ProxyWatcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	collector := &proxyCollector{
+		ProxyWatcherConfig: cfg,
+	}
+	watcher, err := newResourceWatcher(collector, cfg.ResourceWatcherConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &ProxyWatcher{watcher, collector}, nil
+}
+
+// ProxyWatcher is built on top of resourceWatcher to monitor additions
+// and deletions to the set of proxies.
+type ProxyWatcher struct {
+	*resourceWatcher
+	*proxyCollector
```

Comment:
nit: I think that this is slightly overuse of struct embedding that impact readability. The consumer of ProxyWatcher doesn't use any resourceWatcher and proxyCollector methods or fields thus but embedding we are unnecessarily expose resourceWatcher and proxyCollector members and functions.

Reply:
`ProxyWatcher` users call e.g. `Done` from `resourceWatcher` and `GetCurrent` from `proxyCollector`.